integrations: Add webhook code, API endpoint, and tests for Groove.

infinitelooped opened the pull request:
Hi, I have finished writing the code for the integration of Groove to Zulip. Here's a screenshot of possible messages sent:
![screenshot](https://user-images.githubusercontent.com/24468110/34648975-a6553c36-f3d7-11e7-8f13-56d6e9214396.png)

Feedbacks and reviews are appreciated, let me know if you have any requests. Thank you.

roberthoenig commented on zerver/webhooks/groove/tests.py:
````diff
@@ -0,0 +1,124 @@
+        expected_message = u"An open ticket has been assigned to agent@example.com\n" \
+                           u"```quote\n" \
+                           u"**[Ticket #9: Test Subject](https://testteam.groovehq.com/groove_client/tickets/68659446)**\n" \
+                           u"```"
````
You can create prettier multiline strings by using brackets () instead of slashes \, or triple quotes """ instead.

roberthoenig commented on zerver/webhooks/groove/tests.py:
```diff
@@ -0,0 +1,124 @@
+                                          X_GROOVE_EVENT="note_added")
+
+    # This test is for other events (it would
+    # be ignored)
```
Why do we ignore other types of events? We already have a fixture for the `ticket_state_changed` event.

infinitelooped replied:
There are a lot of other events in Groove (there are fifteen in total). However, when Groove is integrated with Slack, we only received eight types of notifications, six of which are identifiable from which event it came from, and the last two are not (I experimented it myself and also consulted to the research note [here](https://github.com/zulip/zulip-gci-submissions/blob/master/webhook-integrations/groove/notes.md)). `ticket_state_changed` is not one of the identified events, and when triggered it does not make a notification message in Slack. Since I'm basing this on the integration in Slack, I treated it as the other types of events and just ignored it.

However, if you wish to include it too as part of the notifications, let me know. There are no messages that can be based on though, so I would write the format of the messages on my own.

infinitelooped commented on zerver/webhooks/groove/tests.py:
```diff
@@ -0,0 +1,124 @@
+    STREAM_NAME = 'groove'
+    URL_TEMPLATE = '/api/v1/external/groove?&api_key={api_key}'
+
+    # This test simulates the notification when
```
Got it, working on these now.

infinitelooped commented on zerver/webhooks/groove/tests.py:
````diff
@@ -0,0 +1,124 @@
+                           u"```quote\n" \
+                           u"**[Ticket #9: Test Subject](https://testteam.groovehq.com/groove_client/tickets/68659446)**\n" \
+                           u"```"
+
````
Sorry, did you mean the blank line on line 31, or the newlines used in the tests?

infinitelooped commented:
Hi, I fixed the comments, but let me know if there's anything more that can be shorten to one line or broken into lines. As per the other events, I already put a comment above about why I ignored them. But again, let me know if I should improvise and implement it too. Thanks.

eeshangarg commented on zerver/webhooks/groove/view.py:
````diff
@@ -0,0 +1,83 @@
+        body += u"({app_ticket_url})**\n{plain_text_body}\n```"
+    else:
+        return json_success()
+
````
@infinitelooped: I would recommend moving all of the if/elif clauses to separate functions, one for each event. So you pass in the payload to each function, which returns the `body` as needed. See [how the GCI webhook does this](https://github.com/zulip/zulip/blob/master/zerver/webhooks/gci/view.py). The reason I am being a bit picky about this is that that kind of organization will be very helpful for adding more configurability to these webhooks in the not-too-distant future!

eeshangarg commented on zerver/webhooks/groove/fixtures/ticket_assigned3.json:
```diff
@@ -0,0 +1,59 @@
+{
```
Could you please give these fixtures names that are more descriptive instead of just numbering them? Just name them in terms of their differences.

infinitelooped commented:
@eeshangarg: I already addressed your issues in the code. I renamed all `ticket_assigned` fixtures to be descriptive about each cases, and put each events into separate functions. Let me know if you need anything more!

On the other hand, the build was failed due to `Page not found: <404 https://pythonhosted.org/Markdown/extensions/api.html>` and I think it's not from my code.

roberthoenig commented on zerver/webhooks/groove/view.py:
```diff
@@ -0,0 +1,102 @@
+    # Other events are ignored.
+    if event in EVENTS_FUNCTION_MAPPER:
+        body = EVENTS_FUNCTION_MAPPER[event](payload)
+    else:
```
It looks like you could simplify this code by removing the else clause and indenting the following "if body is not None:" check.

infinitelooped replied:
Fixed!

showell commented:
Merged!  Thanks @infinitelooped @eeshangarg @roberthoenig
